lib/pure/osproc.nim

```diff
@@ -742,13 +748,31 @@ when defined(Windows) and not defined(useNimRtl):
     result.id = procInfo.dwProcessId
     result.exitFlag = false
 
+  proc closeThreadAndProcessHandle(p: Process) =
+    if p.fThreadHandle != 0:
+      closeHandleCheck(p.fThreadHandle)
+      p.fThreadHandle = 0
+
+    if p.fProcessHandle != 0:
+      closeHandleCheck(p.fProcessHandle)
+      p.fProcessHandle = 0
+
   proc close(p: Process) =
     if poParentStreams notin p.options:
-      discard closeHandle(p.inHandle)
-      discard closeHandle(p.outHandle)
-      discard closeHandle(p.errHandle)
-    discard closeHandle(p.fThreadHandle)
-    discard closeHandle(p.fProcessHandle)
+      if p.inStream == nil:
+        p.inHandle.fileClose()
+      else:
+        # p.inHandle can be already closed via inputStream.
+        p.inStream.close
+
+      # You may NOT close outputStream and errorStream.
+      assert p.outStream == nil or FileHandleStream(p.outStream).handle != INVALID_HANDLE_VALUE
+      assert p.errStream == nil or FileHandleStream(p.errStream).handle != INVALID_HANDLE_VALUE
```

Comment:
`doAssert` because is an important check IMHO.

Reply:
I don't think it should be `doAssert`.
As far as I know, these asserts fails only when `close(p: Process)` was called or outputStream/errorStream was closed before `close(p: Process)` is called.
These errors can be detected if assert was on.
If no asserts failed when assert was on, these error should never happen in -d:danger build.

Reply:
> these error should never happen in -d:danger build.

they could, because a different code path can be executed based on runtime conditions that would only be run in production. Generally, there's no point in skipping some assert if the cost of the check is dwarfed by the cost of the proc it lives in, which is the case here (the syscall would dwarf a simple check)